test/construct.test.did

```diff
@@ -60,7 +60,11 @@ assert blob "DIDL\01\6d\7b\01\00\02\01\02" == "(blob \"\\01\\02\")"         : (v
 assert blob "DIDL\01\6d\00\01\00\00" == "(vec {})"                          : (Vec) "vec: recursive vector";
 assert blob "DIDL\01\6d\00\01\00\02\00\00" == "(vec { vec {}; vec {} })"    : (Vec) "vec: tree";
 assert blob "DIDL\01\6d\00\01\00\02\00\00" == "(vec { vec {}; vec {} })"    : (vec vec opt empty) "vec: non-recursive tree";
-assert blob "DIDL\01\6d\7f\01\00\e8\07"                                     : (vec null) "vec: space bomb";
+assert blob "DIDL\01\6d\7f\01\00\e8\07"                                     : (vec null) "vec null";
+assert blob "DIDL\01\6d\7f\01\00\80\94\eb\dc\03"                           !: (vec opt nat) "space bomb: vec null <: vec opt nat";
+assert blob "DIDL\01\6d\70\01\00\80\94\eb\dc\03"                           !: (opt nat) "space bomb: send vec reserved to opt nat";
+assert blob "DIDL\04\6c\03\01\7f\02\01\03\02\6c\01\01\70\6c\00\6d\00\01\03\80\94\eb\dc\03" !: (vec record {null;record{reserved};record{}}) "space bomb: zero-sized record";
+assert blob "DIDL\02\6d\01\6d\7f\01\00\05\ff\ff\7f\ff\ff\7f\ff\ff\7f\ff\ff\7f\ff\ff\7f"    !: (vec vec null) "space bomb: vec vec null";
```

Comment:
Without the counter, it's hard to reject this test: Each sub `vec null` is of length 2M and takes 3 bytes, the top level vec length is 5. We can easily extend the top level vec length to 2M / 3 = 699050 and still fits in the 2M message limit. The total memory on the receiver end will be way larger than the Wasm heap.

Reply:
Is there a clear and simple criteria underlying this, i.e. when to reject and when to accept? One that makes sense in other implementations as well?

And I'd put them into a new file. This makes it easier for an implementation to treat these tests specially (e.g. test them with a heap size bound, or not at all when just starting to implement Candid.)

Reply:
In the spec, we have `"we limit the total vector length of such zero-sized values in a message to be 2MiB."`. So this message will be rejected when decoding the second `vec null`.